storage-nvmf/src/main/java/org/apache/crail/storage/nvmf/client/NvmfUnalignedWriteFuture.java

```diff
@@ -0,0 +1,183 @@
+			} else {
+				/* Wait for previous end operation to finish */
+				beginBuffer.getFuture().get();
+			}
+			CrailBuffer stagingBuffer = beginBuffer.getBuffer();
+			stagingBuffer.position(offsetInSector(remoteOffset));
+			stagingBuffer.getByteBuffer().put(buffer.getByteBuffer());
+			buffer.limit(oldLimit);
+			stagingBuffer.position(0);
+			beginFuture = endpoint.write(stagingBuffer, blockInfo, alignedRemoteOffset);
+			beginBuffer.setFuture(beginFuture);
+			stagingBuffer.position(offsetInSector(remoteOffset));
+		}
+
+		/* middle */
+		if (isSectorAligned(nextRemoteOffset) && buffer.remaining() >= endpoint.getLBADataSize()) {
+			int oldLimit = buffer.limit();
+			buffer.limit(buffer.position() + (int)floorToSectorSize(buffer.remaining()));
+			int toWrite = buffer.remaining();
+			middleFuture = endpoint.write(buffer, blockInfo, nextRemoteOffset);
+			nextRemoteOffset += toWrite;
+			buffer.position(buffer.limit());
+			buffer.limit(oldLimit);
+		}
+
+		/* end */
+		if (buffer.remaining() > 0) {
+			endBuffer = endpoint.getStagingBufferCache().get(blockInfo.getAddr() + nextRemoteOffset);
+			CrailBuffer stagingBuffer = endBuffer.getBuffer();
+			stagingBuffer.position(0);
+			stagingBuffer.getByteBuffer().put(buffer.getByteBuffer());
+			stagingBuffer.position(0);
+			endFuture = endpoint.write(stagingBuffer, blockInfo, nextRemoteOffset);
+			endBuffer.setFuture(endFuture);
+		}
+	}
+
+	@Override
+	public boolean isSynchronous() {
+		return false;
+	}
+
+	@Override
+	public boolean cancel(boolean b) {
+		return false;
+	}
+
+	@Override
+	public boolean isCancelled() {
+		return false;
+	}
+
+	private static boolean checkIfFutureIsDone(StorageFuture future) {
+		return (future != null && future.isDone()) || future == null;
+	}
+
+	@Override
+	public boolean isDone() {
+		if (beginFuture != null && beginFuture.isDone()) {
+			if (beginBuffer != null) {
+				beginBuffer.put();
+				beginBuffer = null;
+			}
+		}
+		if (endFuture != null && endFuture.isDone()) {
+			if (endBuffer != null) {
+				endBuffer.put();
+				endBuffer = null;
+			}
+		}
+		return beginBuffer == null && checkIfFutureIsDone(middleFuture) && endBuffer == null;
+	}
+
+	@Override
+	public StorageResult get() throws InterruptedException, ExecutionException {
+		try {
+			return get(2, TimeUnit.MINUTES);
+		} catch (TimeoutException e) {
+			throw new ExecutionException(e);
+		}
+	}
+
+	@Override
+	public StorageResult get(long timeout, TimeUnit timeUnit) throws InterruptedException, ExecutionException, TimeoutException {
+		if (!isDone()) {
+			long start = System.nanoTime();
+			long end = start + TimeUnit.NANOSECONDS.convert(timeout, timeUnit);
+			boolean waitTimeOut;
+			do {
+				waitTimeOut = System.nanoTime() > end;
+			} while (!isDone() && !waitTimeOut);
+			if (!isDone() && waitTimeOut) {
+				throw new TimeoutException("poll wait time out!");
+			}
+		}
+		if (beginFuture != null) {
```

Comment:
I don't get this logic here? When would beginFuture (or others) will be null?

Reply:
Any one of the futures can be null. There can be a begin only, end only, begin + middle, begin + middle + end and middle + end operation depending on whether begin or the end sector is unaligned or both. See constructor for the logic.